src/systems/rendering.cpp

```diff
@@ -105,18 +107,51 @@ void MeshRenderSystem::render3D(sp::ecs::Entity e, sp::Transform& transform, Mes
         glActiveTexture(GL_TEXTURE0 + ShaderRegistry::textureIndex(ShaderRegistry::Textures::IlluminationMap));
         mrc.illumination_texture.ptr->bind();
     }
+}
 
-    // Draw
+void drawMesh(MeshRenderComponent& mrc, ShaderRegistry::ScopedShader& shader)
+{
     gl::ScopedVertexAttribArray positions(shader.get().attribute(ShaderRegistry::Attributes::Position));
     gl::ScopedVertexAttribArray texcoords(shader.get().attribute(ShaderRegistry::Attributes::Texcoords));
     gl::ScopedVertexAttribArray normals(shader.get().attribute(ShaderRegistry::Attributes::Normal));
 
     mrc.mesh.ptr->render(positions.get(), texcoords.get(), normals.get());
 
+    // wut iz?
     if (mrc.specular_texture.ptr || mrc.illumination_texture.ptr)
         glActiveTexture(GL_TEXTURE0);
 }
 
+void MeshRenderSystem::update(float delta)
+{
+}
+
+void MeshRenderSystem::render3D(sp::ecs::Entity e, sp::Transform& transform, MeshRenderComponent& mrc)
+{
+    mrc.ensureLoaded();
+
+    auto model_matrix = calculateModelMatrix(
+            transform.getPosition(),
+            transform.getRotation(),
+            mrc);
+
+    auto shader = lookUpShader(mrc);
+    glUniformMatrix4fv(shader.get().uniform(ShaderRegistry::Uniforms::Model), 1, GL_FALSE, glm::value_ptr(model_matrix));
+
+    auto modeldata_matrix = glm::rotate(model_matrix, glm::radians(180.f), {0.f, 0.f, 1.f});
+    modeldata_matrix = glm::scale(modeldata_matrix, glm::vec3{mrc.scale});
+
+    // Lights setup.
+    ShaderRegistry::setupLights(shader.get(), modeldata_matrix);
```

Comment:
I'm not entirely sure I understand why we invert the model for lighting.  I tried rendering with and without the inversion, and while there were differences, they seemed subtle.
Is it to keep things synchronized with other `render3D` rendering?

Reply:
I guess so. This code is old and has seen many revisions from multiple people :)